mm2src/coins/z_coin.rs

```diff
@@ -373,6 +375,75 @@ impl ZCoin {
         }
     }
 
+    /// Gets spendable notes for transaction generation, ensuring previous transactions are confirmed/mined.
+    ///
+    /// # Flow
+    /// 1. Checks for any previous transaction with change that needs confirmation
+    /// 2. If found, waits up to 30 minutes for required confirmations (checking every 15 seconds)
+    /// 3. Once confirmed or if no previous tx, returns ordered list of spendable notes
+    async fn spendable_notes_required_for_tx(&self, required: &BigDecimal) -> MmResult<Vec<SpendableNote>, GenTxError> {
+        let my_balance = self
+            .my_balance()
+            .compat()
+            .await
+            .mm_err(|err| GenTxError::Internal(err.to_string()))?;
+        if my_balance.spendable >= *required {
+            return Ok(self
+                .spendable_notes_ordered()
+                .await
+                .map_err(|err| GenTxError::SpendableNotesError(err.to_string()))?);
+        }
+
+        info!("Tx needs more change");
+        let spendable = my_balance.spendable;
+        let unspendable = my_balance.unspendable;
+        let total_balance = &spendable + &unspendable;
+
+        if &total_balance < required {
+            return MmError::err(GenTxError::InsufficientBalance {
+                coin: self.ticker().to_string(),
+                available: spendable,
+                required: required.clone(),
+            });
+        }
+
+        let amount_needed = required - spendable;
+        let available_changes = self.z_fields.change_note_db.load_all_notes().await?;
+        let mut amount_collected = BigDecimal::default();
+
+        for note in available_changes {
+            amount_collected += big_decimal_from_sat(note.change as i64, self.decimals());
+            // Wait for tx confirmation
+            // Wait up to 30 minutes for confirmations with 15 sec check interval
+            // (probably will be changed)
+            let wait_until = now_sec() + (30 * 60);
```

Comment:
I think maybe it's better not to do waiting inside the KDF and just delegate this to the User: simply return unconfirmed balance to the GUI (GUI may display unconfirmed balance and a warning that confirmation is needed)

Reply:
the is the same as gui making sure that there's no more than one ARRR swap running..right?

Reply:
I still think that we should not wait for confirmation inside a low level function like gen_tx, used both for withdraw and swaps.
I think, for withdraw we should wait inside the KDF but just return unconfirmed (unspendable) balance to the GUI so the User can wait for it to be confirmed, watching the balance status.
For swaps, if we detected unspendable balance needed to continue the swap (make payment), we should wait for it inside the workflow step.

Reply:
the code you are commenting on is outdated

Reply:
resolved due to the outdated code